Begin implementing RemoteDOMWindow properties

achristensen07 opened the pull request:
#### c44dcfa19a644523ebdc60d75d94dbdcf8a71e4b
<pre>
Begin implementing RemoteDOMWindow properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=262327">https://bugs.webkit.org/show_bug.cgi?id=262327</a>
rdar://116199379

Reviewed by Pascoe.

A cross-origin iframe has some properties that can be queried without site isolation,
and these properties need to continue to work with site isolation enabled.

length is the number of subframes, which we can get from the frame tree.
parent goes up in the frame tree by one frame
top goes up in the frame tree all the way to ... the top

Scoped children checks work a little differently with RemoteFrames and RemoteDOMWindows
because we don&apos;t have a Document or a TreeScope from that Frame.  For now, consider these
in scope until removed by an incoming IPC message.  This definition may need some refining
later, but this makes these cases work as they used to without site isolation enabled.

In order to get the test to finish successfully, I had to correct the conditions under
which WebPage::DidFinishLoadInAnotherProcess would be sent.  Instead of checking if the
parent frame has a RemotePageProxy, we need to check if the frame is in a different process
than its parent.  This makes the test finish successfully when the grandchild
is in the same process as the main frame (so it has no RemotePageProxy) but its parent
is not.  They are in different processes, so IPC is required to inform of the frame
load completion.

* LayoutTests/http/tests/site-isolation/resources/window-properties-child.html: Added.
* LayoutTests/http/tests/site-isolation/resources/window-properties-grandchild.html: Added.
* LayoutTests/http/tests/site-isolation/window-properties-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/window-properties.html: Added.
* Source/WebCore/page/FrameTree.cpp:
(WebCore::inScope):
* Source/WebCore/page/RemoteDOMWindow.cpp:
(WebCore::RemoteDOMWindow::focus):
(WebCore::RemoteDOMWindow::length const):
(WebCore::RemoteDOMWindow::top const):
(WebCore::RemoteDOMWindow::parent const):

Canonical link: <a href="https://commits.webkit.org/268688@main">https://commits.webkit.org/268688@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/96a701dbc79fdbf94b8c187330a5df88bde96027

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20332 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20761 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21404 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22230 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18970 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20563 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24019 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20934 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20392 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20551 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20445 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17677 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23080 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17612 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18483 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24786 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18690 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18659 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22720 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19239 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16340 "10 flakes 2 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18451 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/18319 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22788 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2526 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19064 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/WebCore/page/FrameTree.cpp:
```diff
@@ -126,7 +126,7 @@ static bool inScope(Frame& frame, TreeScope& scope)
     auto* localFrame = dynamicDowncast<LocalFrame>(frame);
     if (!localFrame)
-        return false;
+        return true;
```
Is it true that any remote frames passed in here would be inScope?

achristensen07 replied:
I think so.  The scope seems to be used to prevent detached frames from being included.  I filed rdar://116268930 for future investigation.

achristensen07 committed:
Begin implementing RemoteDOMWindow properties
https://bugs.webkit.org/show_bug.cgi?id=262327
rdar://116199379

Reviewed by Pascoe.

A cross-origin iframe has some properties that can be queried without site isolation,
and these properties need to continue to work with site isolation enabled.

length is the number of subframes, which we can get from the frame tree.
parent goes up in the frame tree by one frame
top goes up in the frame tree all the way to ... the top

Scoped children checks work a little differently with RemoteFrames and RemoteDOMWindows
because we don't have a Document or a TreeScope from that Frame.  For now, consider these
in scope until removed by an incoming IPC message.  This definition may need some refining
later, but this makes these cases work as they used to without site isolation enabled.

In order to get the test to finish successfully, I had to correct the conditions under
which WebPage::DidFinishLoadInAnotherProcess would be sent.  Instead of checking if the
parent frame has a RemotePageProxy, we need to check if the frame is in a different process
than its parent.  This makes the test finish successfully when the grandchild
is in the same process as the main frame (so it has no RemotePageProxy) but its parent
is not.  They are in different processes, so IPC is required to inform of the frame
load completion.

* LayoutTests/http/tests/site-isolation/resources/window-properties-child.html: Added.
* LayoutTests/http/tests/site-isolation/resources/window-properties-grandchild.html: Added.
* LayoutTests/http/tests/site-isolation/window-properties-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/window-properties.html: Added.
* Source/WebCore/page/FrameTree.cpp:
(WebCore::inScope):
* Source/WebCore/page/RemoteDOMWindow.cpp:
(WebCore::RemoteDOMWindow::focus):
(WebCore::RemoteDOMWindow::length const):
(WebCore::RemoteDOMWindow::top const):
(WebCore::RemoteDOMWindow::parent const):

Canonical link: https://commits.webkit.org/268688@main
